Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPICookiesCocoa.mm

```diff
@@ -0,0 +1,265 @@
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#if !__has_feature(objc_arc)
+#error This file requires ARC. Add the "-fobjc-arc" compiler flag for this file.
+#endif
+
+#import "config.h"
+#import "WebExtensionAPICookies.h"
+
+#if ENABLE(WK_WEB_EXTENSIONS)
+
+#import "CocoaHelpers.h"
+#import "Logging.h"
+#import "MessageSenderInlines.h"
+#import "WebExtensionAPINamespace.h"
+#import "WebExtensionContextMessages.h"
+#import "WebExtensionContextProxy.h"
+#import "WebExtensionUtilities.h"
+#import "WebProcess.h"
+
+namespace WebKit {
+
+static NSString * const domainKey = @"domain";
+static NSString * const expirationDateKey = @"expirationDate";
+static NSString * const httpOnlyKey = @"httpOnly";
+static NSString * const nameKey = @"name";
+static NSString * const partitionKeyKey = @"partitionKey";
+static NSString * const pathKey = @"path";
+static NSString * const sameSiteKey = @"sameSite";
+static NSString * const secureKey = @"secure";
+static NSString * const sessionKey = @"session";
+static NSString * const storeIdKey = @"storeId";
+static NSString * const topLevelSiteKey = @"topLevelSite";
+static NSString * const urlKey = @"url";
+static NSString * const valueKey = @"value";
+
+void WebExtensionAPICookies::get(NSDictionary *details, Ref<WebExtensionCallbackHandler>&& callback, NSString **outExceptionString)
+{
+    // Documentation: https://developer.mozilla.org/docs/Mozilla/Add-ons/WebExtensions/API/cookies/get
+
+    static NSArray<NSString *> *requiredKeys = @[
+        nameKey,
+        urlKey,
+    ];
+
+    static NSDictionary<NSString *, id> *types = @{
+        nameKey: NSString.class,
+        partitionKeyKey: NSDictionary.class,
+        storeIdKey: NSString.class,
+        urlKey: NSString.class,
+    };
+
+    if (!validateDictionary(details, @"details", requiredKeys, types, outExceptionString))
+        return;
+
+    if (NSDictionary *partitionKey = details[partitionKeyKey]) {
+        if (!validateDictionary(partitionKey, partitionKeyKey, nil, @{ topLevelSiteKey: NSString.class }, outExceptionString))
+            return;
+    }
+
+    if (NSString *name = details[nameKey]; !name.length) {
+        *outExceptionString = toErrorString(nil, nameKey, @"it must not be empty");
+        return;
+    }
+
+    if (NSString *urlString = details[urlKey]) {
+        if (!urlString.length) {
```

Comment:
could combine into one line like you did above

Reply:
In this case there are additional checks, so it might not flow as cleanly.